### PR TITLE
Improve many doc query builder api

### LINKED ISSUE #31: Changes to Query and Document builder
I think our current Query and Document builders could be a bit more consistent. 

We currently have 8 public overloaded `queryXX` methods in `GetManyDocumentsQueryBuilder` and 6 in `ViewQueryBuilder`, most of which exist to address the fact that they have diff return types, even though the differences between them are typically http query params. 

This issue proposes an API change, where `includeDocs`, `reduce`, `allowMissing`,`byType`... do not require separate `queryXX` methods but are instead a part of the builder config. All `queryXX` methods would thus be replaced for a single `query(ids)` method, which under the hood might use different Query strategy impls based on the configs of the builder and a prior `build` step.

**e.g.**
_Current_
 `Documents.getMany.descending.limit(10).queryIncludeDocsAllowMissing(ids)`
 `Documents.getMany.descending.limit(10).queryIncludeDocs(ids)`
 `Query.view.limit(10).queryWithReduce`

_Proposed_
`Documents.getMany.descending.limit(10).allowMissing.includeDocs.build.query(ids)`
`Documents.getMany.descending.limit(10).includeDocs.query`
`Query.view.limit(10).reduce.build.query`

### PR DESCRIPTION
**Pull request by ermyas**
fixes #31

### CODE REVIEW

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 1]
```diff
@@ -25,92 +25,135 @@ import upickle.default.write
...
+  }
+
+  def endKeyDocId(
+      endKeyDocId: String): GetManyDocumentsQueryBuilder[ID, AM, BT] = {
+    set[ID, AM, BT]("endkey_docid", endKeyDocId)
   }
 
-  def conflicts(conflicts: Boolean = true): GetManyDocumentsQueryBuilder = {
-    set("conflicts", conflicts)
+  def includeDocs[D: R]: GetManyDocumentsQueryBuilder[IncludeDoc[D], AM, BT] = {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: What do you think of having a boolean argument here and for `allowMissing`, so that it can be turned off?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 2]
```diff
@@ -25,92 +25,135 @@ import upickle.default.write
...
 
-  def endKey[K: W](endKey: K): GetManyDocumentsQueryBuilder = {
-    set("endkey", write(endKey))
+  def withIds(ids: Seq[String]): GetManyDocumentsQueryBuilder[ID, AM, BT] = {
+    set[ID, AM, BT](params, ids, view)
   }
 
-  def endKeyDocId(endKeyDocId: String): GetManyDocumentsQueryBuilder = {
-    set("endkey_docid", endKeyDocId)
+  def byType[K: R, V: R, D: R](_view: CouchView):
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: Should the argument be called `view`? `_view` is unused at the moment here.

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 3]
```diff
@@ -25,92 +25,135 @@ import upickle.default.write
...
+
+sealed trait MissingIdsInQuery
+trait MissingAllowed extends MissingIdsInQuery
+trait MissingNotAllowed extends MissingIdsInQuery
+
+sealed trait DocType
+abstract class ForDocType[D: R, K: R, V: R] extends DocType
+abstract class AnyDocType extends DocType
+
+case class GetManyDocumentsQueryBuilder[ID <: DocsInResult, AM <: MissingIdsInQuery, BT <: DocType](
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: Should this case class be private?

File: src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    CouchView(
+      map =
+          """
+            |function(doc) {
+            | emit([doc.kind, doc._id], doc._id);
+            |}
+          """.stripMargin)
+  }
+
+  def postQuery[Q: R](
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala]: Why is this called `postQuery` if it's making a get request?

File: src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.ibm.couchdb.core.Client
+import com.ibm.couchdb.{CouchView, Req}
+import org.http4s.Status
+import upickle.default.Aliases.{R, W}
+import upickle.default._
+
+import scalaz.concurrent.Task
+
+object QueryUtils {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala]: Is this a good way to group these things? Usually, `xUtils` become containers of lots of unrelated stuff over time. Is it worth splitting it into more specific classes / traits?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 6]
```diff
@@ -160,13 +205,69 @@ case class GetManyDocumentsQueryBuilder(
...
+  trait BuildReady[T] {
+    def build: T
+  }
+
+  def apply(client: Client, db: String, typeMapping: TypeMapping):
+  GetManyDocumentsQueryBuilder[ExcludeDocs, MissingNotAllowed, AnyDocType] =
+    new GetManyDocumentsQueryBuilder[ExcludeDocs, MissingNotAllowed, AnyDocType](
+      client, db, typeMapping)
+
+  implicit def buildBasic(
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: I didn't think enough about it, but is there a way to do this without implicits?

Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: I didn't think enough about it, but is there a way to do this without implicits?

File: scalastyle-config.xml [hunk 7]
```diff
@@ -109,7 +109,7 @@
  </check>
  <check level="warning" class="org.scalastyle.scalariform.NumberOfMethodsInTypeChecker" enabled="true">
   <parameters>
-   <parameter name="maxMethods"><![CDATA[30]]></parameter>
+   <parameter name="maxMethods"><![CDATA[35]]></parameter>
```
Comment (beloglazov) [inline scalastyle-config.xml]: If this is for `GetManyDocumentsQueryBuilder`, let's add a todo to change it back when the deprecated methods are removed.

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 2 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: good catch

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 1 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: I initially thought it might be a bit redundant since these are always turned off by default `GetManyDocumentsQueryBuilder[ExcludeDocs, MissingNotAllowed, AnyDocType]` and thus require explicit calls to turn these on, which makes scenario that you would want turn it backoff a bit less likely? ... but could still be useful I guess ... adding overloaded methods for this so as to still have the convenience of not having to specify braces :-)

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 3 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: were you referring to the constructor? if so yes... but probably not the case class itself since its type is required outside this class

File: src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala (hunk 4 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala]: good catch

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 6 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: tried a couple of different methods but they were messy and brittle. This approach to a type-safe builder is discussed [here](http://blog.rafaelferreira.net/2008/07/type-safe-builder-pattern-in-scala.html) 

open to suggestions though :-)

File: src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala [hunk 8]
```diff
@@ -16,39 +16,57 @@
...
+  }
+}
 
-  def postQuery[Q: R](
-      client: Client,
-      db: String,
-      url: String,
-      params: Map[String, String]): Task[Q] = {
-    client.get[Q](url, Status.Ok, params.toSeq)
+case class QueryIncludeDocs[C: R, D: R](
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala]: This class seems to be exactly the same as QueryBasic, except or the type parameter D.

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 9]
```diff
@@ -25,92 +25,143 @@ import upickle.default.write
 import scala.reflect.ClassTag
 import scalaz.concurrent.Task
 
-case class GetManyDocumentsQueryBuilder(
+sealed trait DocsInResult
+abstract class IncludeDoc[D: R] extends DocsInResult
+abstract class ExcludeDocs extends DocsInResult
+
+sealed trait MissingIdsInQuery
+trait MissingAllowed extends MissingIdsInQuery
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:33]: Why are these traits and the other ones are abstract classes?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 10]
```diff
@@ -139,16 +192,16 @@ case class GetManyDocumentsQueryBuilder(
       "It uses temporary views to perform type based filters and is inefficient. " +
       "Instead, create a permanent view for type based filtering and use the " +
       "`queryByTypeIncludeDocs[K, V, D: R] (typeFilterView: CouchView) method.")
-    queryByTypeIncludeDocs[(String, String), String, D](tempTypeFilterView)
+    queryByTypeIncludeDocs[(String, String), String, D](QueryUtils.tempTypeFilterView)
   }
 
   def queryIncludeDocs[D: R](ids: Seq[String]): Task[CouchDocs[String, CouchDocRev, D]] = {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:219]: Also deprecated?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 11]
```diff
@@ -139,16 +192,16 @@ case class GetManyDocumentsQueryBuilder(
...
-    queryByIds[CouchDocs[String, CouchDocRev, D]](ids, includeDocs().params)
+    queryByIds[CouchDocs[String, CouchDocRev, D]](ids, includeDocs[D].params)
   }
 
   def queryIncludeDocsAllowMissing[D: R](
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:225]: Deprecated?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 1 above)
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: 👍

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 12]
```diff
@@ -160,13 +213,69 @@ case class GetManyDocumentsQueryBuilder(
...
       Res.Error("not_found", "No IDs specified").toTask
     else
-      queryByIds[String, Q](client, db, s"/$db/_all_docs", ids, ps)
+      QueryUtils.queryByIds[String, Q](client, db, s"/$db/_all_docs", ids, ps)
+  }
+}
+
+object GetManyDocumentsQueryBuilder {
+
+  trait BuildReady[T] {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: Adding implementations of this trait for the query strategies could simplify the implicits (especially if QueryBasic and QueryIncludeDocs get merged).

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 6 above)
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: After thinking about it more and considering that we want to have different return types based on the configuration, this approach looks nice. There is quite a bit of boilerplate, maybe some of it can be removed by adding implementations of BuildReady.

File: src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+          """
+            |function(doc) {
+            | emit([doc.kind, doc._id], doc._id);
+            |}
+          """.stripMargin)
+  }
+
+  def query[Q: R](
+      client: Client,
+      db: String,
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala]: db is unused.

File: src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+      url: String,
+      ids: Seq[K],
+      params: Map[String, String]): Task[Q] = {
+    postQuery[Req.DocKeys[K], Q](client, db, url, Req.DocKeys(ids), params)
+  }
+
+  def postQuery[B: W, Q: R](
+      client: Client,
+      db: String,
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala]: db is unused.

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 6 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: 👍

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 9 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:33]: those that require type params have to be abstract classes I guess since traits can't have type params with context bounds
 `MissingAllowed` and `MissingNotAllowed` don't require type params. 
Were you thinking of changing this for consistency?
Admitedly `ExcludeDocs` and `AnyDocs` can also be a traits

File: src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala (hunk 14 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryUtils.scala]: not sure why intellij code inspection did not pick this up. Would have been good to have a rule in scalastyle checker for checking unused params

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 15]
```diff
@@ -232,55 +232,51 @@ BT <: DocType] private(
...
-        builder.client, builder.db, builder.view, builder.typeMappings)
-    }
+      builder: MDBuilder[IncludeDoc[D], MissingNotAllowed, ForDocType[K, V, D]])
+      (implicit tag: ClassTag[D], kw: W[K]): ByTypeBuilder[K, V, D] = {
+    new ByTypeBuilder[K, V, D](builder)
   }
+
+  def apply(client: Client, db: String, typeMapping: TypeMapping):
+  MDBuilder[ExcludeDocs, MissingNotAllowed, AnyDocType] =
+    new MDBuilder[ExcludeDocs, MissingNotAllowed, AnyDocType](client, db, typeMapping)
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: Are these type annotations required here? Can they be inferred based on the return type?

File: src/main/scala/com/ibm/couchdb/api/builders/QueryOps.scala [hunk 16]
```diff
@@ -24,17 +24,7 @@ import upickle.default._
...
-  lazy val tempTypeFilterView: CouchView = {
-    CouchView(
-      map =
-          """
-            |function(doc) {
-            | emit([doc.kind, doc._id], doc._id);
-            |}
-          """.stripMargin)
-  }
+trait QueryOps {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryOps.scala]: Can this be a class that takes `client` as a constructor argument?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 17]
```diff
@@ -245,7 +256,11 @@ object GetManyDocumentsQueryBuilder {
       builder: MDBuilder[IncludeDoc[D], MissingNotAllowed, ForDocType[K, V, D]])
       (implicit tag: ClassTag[D], kw: W[K]) {
     def build: QueryByType[K, V, D] =
-      new QueryByType[K, V, D](builder.client, builder.db, builder.view, builder.typeMappings)
+      builder.view match {
+        case Some(view) => new QueryByType[K, V, D](
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: What happens if `view` is `None`?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 18]
```diff
@@ -43,87 +57,137 @@ case class GetManyDocumentsQueryBuilder(
...
+  def endKey[K: W](endKey: K): GetManyDocumentsQueryBuilder[ID, AM, BT] = {
+    set[ID, AM, BT]("endkey", write(endKey))
+  }
+
+  def endKeyDocId(
+      endKeyDocId: String): GetManyDocumentsQueryBuilder[ID, AM, BT] = {
+    set[ID, AM, BT]("endkey_docid", endKeyDocId)
+  }
+
+  def includeDocs[D: R]: GetManyDocumentsQueryBuilder[IncludeDocs[D], AM, BT] = {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:81]: Let's maybe keep the argument here for consistency with the other methods, by default set to true? Will this still allow the method to be used without parentheses?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 19]
```diff
@@ -160,13 +230,68 @@ case class GetManyDocumentsQueryBuilder(
...
   }
+
+  case class ByTypeBuilder[K: R, V: R, D: R](
+      builder: MDBuilder[IncludeDocs[D], MissingNotAllowed, ForDocType[K, V, D]])
+      (implicit tag: ClassTag[D], kw: W[K]) {
+    def build: QueryByType[K, V, D] =
+      builder.view match {
+        case Some(view) => new QueryByType[K, V, D](
+          builder.client, builder.db, view,
+          builder.typeMappings)
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: Unnecessary line break.

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 20]
```diff
@@ -160,13 +230,68 @@ case class GetManyDocumentsQueryBuilder(
...
+
+  implicit def buildByTypeIncludeDocs[K: R, V: R, D: R](
+      builder: MDBuilder[IncludeDocs[D], MissingNotAllowed, ForDocType[K, V, D]])
+      (implicit tag: ClassTag[D], kw: W[K]): ByTypeBuilder[K, V, D] = {
+    new ByTypeBuilder[K, V, D](builder)
+  }
+
+  def apply(client: Client, db: String, typeMapping: TypeMapping):
+  MDBuilder[ExcludeDocs, MissingNotAllowed, AnyDocType] =
+    new MDBuilder[ExcludeDocs, MissingNotAllowed, AnyDocType](client, db, typeMapping)
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: Are these type annotations required here? Can they be inferred based on the return type?

File: src/main/scala/com/ibm/couchdb/api/builders/QueryOps.scala [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.ibm.couchdb.core.Client
+import com.ibm.couchdb.{CouchView, Req}
+import org.http4s.Status
+import upickle.default.Aliases.{R, W}
+import upickle.default._
+
+import scalaz.concurrent.Task
+
+trait QueryOps {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryOps.scala]: Can this be a class that takes `client` as a constructor argument?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 22]
```diff
@@ -160,13 +230,68 @@ case class GetManyDocumentsQueryBuilder(
...
+    def build: QueryBasic[T] =
+      new QueryBasic[T](builder.client, builder.db, builder.params, builder.ids)
   }
+
+  case class ByTypeBuilder[K: R, V: R, D: R](
+      builder: MDBuilder[IncludeDocs[D], MissingNotAllowed, ForDocType[K, V, D]])
+      (implicit tag: ClassTag[D], kw: W[K]) {
+    def build: QueryByType[K, V, D] =
+      builder.view match {
+        case Some(view) => new QueryByType[K, V, D](
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: What happens if `view` is `None`?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 18 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:81]: - not sure how this would work since the value of the parameter for include docs would influence the return type `IncludeDocs[D]` or `ExcludeDocs` 
- no you would have to use parentheses.
- the type parameter `D` becomes unnecessary for includeDocs(false) as well but this isn't a show stopper

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 20 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: explicit specification of return types for public methods is part of our style checker's configuration. did you want to override this?

File: src/main/scala/com/ibm/couchdb/api/builders/QueryOps.scala (hunk 21 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryOps.scala]: it could, but I think it gives us more flexibility as a trait ... clients can extend another class but still include the functionality of this trait etc.

however, to reduce the boiler plate of the query methods, we can add an abstract declaration for a client ( `def client: Client`) within the trait that the dependent classes specify/override

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 22 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: this should not happen :-) 
we can add a case for `None` that uses the defined temporary view, but this does not sound right to me since it is covering a more sinister flaw in the builder logic .. best to fail fast if such a thing occurs

File: src/main/scala/com/ibm/couchdb/api/builders/QueryOps.scala (hunk 21 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryOps.scala]: unless you had composition in mind and not inheritance, once `QueryOps` is a class? which sounds reasonable :-)

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 20 above)
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: I meant `new MDBuilder(client, db, typeMapping)`

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 18 above)
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:81]: In that case, maybe add an overload similar to `allowMissing`? It would have to have an `if` statement and return `IncludeDocs[D]` or `ExcludeDocs` based on the argument. The reason I asked for this is that I tried keeping the method names match the parameter names in the CouchDB API, this makes it easier for the users to find the method. With the rest of the setters you can just pass a `false` to disable the option. If we don't have it here, the users would have to guess that they need to call `excludeDocs` instead of `includeDocs(false)` (which is like it's done for the other options). However, I'm not sure if it's really necessary (having multiple ways to do the same thing is not very good either), what do you think? Maybe to keep it consistent, remove `excludeDocs`?

Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:81]: yea the problem I was trying to get at with the return type of `includeDocs[D](Boolean)` is that its return type would have to be `GetManyDocumentsQueryBuilder[_, AM, BT]` to capture both cases of `IncludeDocs[D]` and `ExcludeDocs`? This creates problems with implicit resolution

but fair point about how such a method would make a more consistent way to supporting document exclusion

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 20 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: ah 👍

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 23]
```diff
@@ -160,13 +232,69 @@ case class GetManyDocumentsQueryBuilder(
...
+
+object GetManyDocumentsQueryBuilder {
+
+  type MDBuilder[ID <: DocsInResult, AM <: MissingIdsInQuery, BT <: DocType] =
+  GetManyDocumentsQueryBuilder[ID, AM, BT]
+
+  case class Builder[T: R, ID <: DocsInResult, AM <: MissingIdsInQuery]
+  (builder: MDBuilder[ID, AM, AnyDocType]) {
+    def build: QueryBasic[T] =
+      new QueryBasic[T](builder.client, builder.db, builder.params, builder.ids)
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: Can the type parameter be removed here as well?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 24]
```diff
@@ -160,13 +232,69 @@ case class GetManyDocumentsQueryBuilder(
...
+  }
+
+  case class ByTypeBuilder[K: R, V: R, D: R](
+      builder: MDBuilder[IncludeDocs[D], MissingNotAllowed, ForDocType[K, V, D]])
+      (implicit tag: ClassTag[D], kw: W[K]) {
+    def build: QueryByType[K, V, D] = {
+      val view = builder.view.getOrElse(builder.tempTypeFilterView)
+      QueryByType[K, V, D](builder.client, builder.db, view, builder.typeMappings)
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: Same with the type parameters here?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala (hunk 18 above)
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala:81]: If you set the return type to `GetManyDocumentsQueryBuilder[ID, AM, BT]`, the implicits won't resolve either?

File: src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala [hunk 25]
```diff
@@ -43,87 +59,136 @@ case class GetManyDocumentsQueryBuilder(
...
+
+  def excludeDocs: GetManyDocumentsQueryBuilder[ExcludeDocs, AM, BT] = {
+    set("include_docs", false)
+  }
+
+  def allowMissing: GetManyDocumentsQueryBuilder[ID, MissingAllowed, BT] = {
+    allowMissing(allow = true)
+  }
+
+  def allowMissing(allow: Boolean): GetManyDocumentsQueryBuilder[ID, MissingAllowed, BT] = {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/GetManyDocumentsQueryBuilder.scala]: This doesn't look right, `allow` is unused, and the return type is always `MissingAllowed`.

**Comment by beloglazov** [comment]
Awesome work, LGTM!
